### PR TITLE
Added support for SRP auth

### PR DESCRIPTION
**Pull request by armicron**
#1 
I've tested with `python2.7` and with a user that don't require a password change on auth.
Please, review and comment.

### CODE REVIEW

File: warrant/aws_srp.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    :param {Long integer} A Large A value.
+    :param {Long integer} B Server B value.
+    :return {Long integer} Computed U value.
+    """
+    UHexHash = hexHash(padHex(A) + padHex(B))
+    finalU = hex_to_long(UHexHash)
+    return finalU
+
+
+class AwsSrp:
```
Comment (bjinwright) [inline warrant/aws_srp.py]: class AWSSRP(object):

File: warrant/aws_srp.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        self.username = username
+        self.password = password
+        self.pool_id = pool_id
+        self.N = hex_to_long(n_hex)
+        self.g = hex_to_long(g_hex)
+        self.k = hex_to_long(hexHash('00' + n_hex + '0' + g_hex))
+        self.smallAValue = self.generateRandomSmallA()
+        self.largeAValue = self.calculateA()
+
+    def generateRandomSmallA(self):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def generate_random_small_a(self):

File: warrant/aws_srp.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        """
+        helper function to generate a random big integer
+        :return {Long integer} a random value.
+        """
+        randomBigInt = get_random(128)
+        smallABigInt = randomBigInt % self.N
+        return smallABigInt
+
+    def calculateA(self):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def calculate_a(self):

File: warrant/aws_srp.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        :param {Long integer} a Randomly generated small A.
+        :return {Long integer} Computed large A.
+        """
+        A = pow(self.g, self.smallAValue, self.N)
+        # safety check
+        if (A % self.N) == 0:
+            raise ValueError('Safety check for A failed')
+        return A
+
+    def getPasswordAuthenticationKey(self, username, password, serverBValue, salt):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def get_password_auth_key(self, username, password, server_b_value, salt):

File: warrant/aws_srp.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    :return {Buffer} Strong key material.
+    @private
+    """
+    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
+    infoBitsUpdate = infoBits + bytearray(chr(1))
+    hmac_hash = hmac.new(prk, infoBitsUpdate, hashlib.sha256).digest()
+    return hmac_hash[:16]
+
+
+def calculateU(A, B):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def calculate_u(A,B):

File: warrant/aws_srp.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    else:
+        hashStr = bigInt
+    if len(hashStr) % 2 == 1:
+        hashStr = '0%s' % hashStr
+    elif hashStr[0] in '89ABCDEFabcdef':
+        hashStr = '00%s' % hashStr
+    return hashStr
+
+
+def computehkdf(ikm, salt):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def compute_hkdf(ikm, salt):

File: warrant/aws_srp.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+def long_to_hex(long_num):
+    return '%x' % long_num
+
+
+def get_random(nbytes):
+    random_hex = os.urandom(nbytes).encode('hex')
+    return hex_to_long(random_hex)
+
+
+def padHex(bigInt):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def pad_hex(big_int):

File: warrant/aws_srp.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+infoBits = bytearray('Caldera Derived Key', 'utf-8')
+
+
+def hash_sha256(buf):
+    """AuthenticationHelper.hash"""
+    a = hashlib.sha256(buf).hexdigest()
+    return (64 - len(a)) * '0' + a
+
+
+def hexHash(hexStr):
```
Comment (bjinwright) [inline warrant/aws_srp.py]: def hex_hash(hex_str):

File: warrant/aws_srp.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        'EF9519B3CD3A431B302B0A6DF25F14374FE1356D6D51C245' + 'E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED' + \
+        'EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3D' + 'C2007CB8A163BF0598DA48361C55D39A69163FA8FD24CF5F' + \
+        '83655D23DCA3AD961C62F356208552BB9ED529077096966D' + '670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B' + \
+        'E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9' + 'DE2BCBF6955817183995497CEA956AE515D2261898FA0510' + \
+        '15728E5A8AAAC42DAD33170D04507A33A85521ABDF1CBA64' + 'ECFB850458DBEF0A8AEA71575D060C7DB3970F85A6E1E4C7' + \
+        'ABF5AE8CDB0933D71E8C94E04A25619DCEE3D2261AD2EE6B' + 'F12FFA06D98A0864D87602733EC86A64521F2B18177B200C' + \
+        'BBE117577A615D6C770988C0BAD946E208E24FA074E5AB31' + '43DB5BFCE0FD108E4B82D120A93AD2CAFFFFFFFFFFFFFFFF'
+# https://github.com/aws/amazon-cognito-identity-js/blob/master/src/AuthenticationHelper.js#L49
+g_hex = '2'
+infoBits = bytearray('Caldera Derived Key', 'utf-8')
```
Comment (bjinwright) [inline warrant/aws_srp.py]: info_bits = bytearray('Caldera Derived Key', 'utf-8')

**Comment by armicron** [comment]
Tested with python3.4 and requested changes are made.

File: warrant/aws_srp.py [hunk 10]
```diff
@@ -165,3 +166,19 @@ def process_challenge(self, challenge_parameters):
                 "USERNAME": user_id_for_srp,
                 "PASSWORD_CLAIM_SECRET_BLOCK": secret_block_b64,
                 "PASSWORD_CLAIM_SIGNATURE": signature_string.decode('utf-8')}
+
+    def authenticate_user(self):
+        self.client = boto3.client('cognito-idp')
```
Comment (armicron) [inline warrant/aws_srp.py]: I don't know, maybe I should pass the client from Cognito?

Comment (bjinwright) [inline warrant/aws_srp.py]: Yeah I see what you're saying

**Comment by mcrowson** [comment]
bad ass, thanks @armicron

File: warrant/aws_srp.py (hunk 10 above)
Comment (bjinwright) [inline warrant/aws_srp.py]: Ok I have a solution. Add client=None as a kwarg in __init__ and add self.client = client. Also add client=None as a kwarg in the authenticate_user method. Inside the method add an if statement that either uses self.client or the local client kwarg.

**Pull request merged**